src/os/src/detail/osTime.c

```diff
@@ -622,56 +628,22 @@ int64_t taosTimeTruncate(int64_t t, const SInterval* pInterval, int32_t precisio
 
     if (pInterval->slidingUnit == 'y') {
       tm.tm_mon = 0;
-      tm.tm_year = (int)(tm.tm_year / pInterval->sliding * pInterval->sliding);
+      tm.tm_year -= (int)(tm.tm_year%pInterval->sliding);
     } else {
       int mon = tm.tm_year * 12 + tm.tm_mon;
-      mon = (int)(mon / pInterval->sliding * pInterval->sliding);
+      mon -= (int)(mon % pInterval->sliding);
       tm.tm_year = mon / 12;
       tm.tm_mon = mon % 12;
     }
 
     start = (int64_t)(mktime(&tm) * TSDB_TICK_PER_SECOND(precision));
   } else {
-    int64_t delta = t - pInterval->interval;
-    int32_t factor = (delta >= 0) ? 1 : -1;
-
-    start = (delta / pInterval->sliding + factor) * pInterval->sliding;
-
-    if (pInterval->intervalUnit == 'd' || pInterval->intervalUnit == 'w') {
-     /*
-      * here we revised the start time of day according to the local time zone,
-      * but in case of DST, the start time of one day need to be dynamically decided.
-      */
-      // todo refactor to extract function that is available for Linux/Windows/Mac platform
-  #if defined(WINDOWS) && _MSC_VER >= 1900
-      // see https://docs.microsoft.com/en-us/cpp/c-runtime-library/daylight-dstbias-timezone-and-tzname?view=vs-2019
-      int64_t timezone = _timezone;
-      int32_t daylight = _daylight;
-      char**  tzname = _tzname;
-  #endif
-
-      start += (int64_t)(timezone * TSDB_TICK_PER_SECOND(precision));
-    }
-
-    int64_t end = 0;
-
-    // not enough time range
-    if (start < 0 || INT64_MAX - start > pInterval->interval - 1) {
-      end = start + pInterval->interval - 1;
-
-      while(end < t && ((start + pInterval->sliding) <= INT64_MAX)) { // move forward to the correct time window
-        start += pInterval->sliding;
-
-        if (start < 0 || INT64_MAX - start > pInterval->interval - 1) {
-          end = start + pInterval->interval - 1;
-        } else {
-          end = INT64_MAX;
-          break;
-        }
-      }
-    } else {
-      end = INT64_MAX;
-    }
+    // To avoid the overly complicated effect of DST on size of sliding windows, the DST(daylight saving time) is
+    // better to be respected by users' input and output for display only, maintaining the status quo.
+    // In this way, the size of sliding windows keeps unchanging during each cycle of processing of requests.
+    start = t - pInterval->interval + pInterval->sliding;
+    int64_t tz_offset = taosTimeTzOffset(pInterval->intervalUnit, precision);
+    start -= (start+tz_offset)%pInterval->sliding;
```

Comment:
Truncate by subtracting the remainder within aligned intervals.

Reply:
The effect of code keeps exactly the same as old, while readability of the code improves significantly.